core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1435,19 +1452,22 @@ class Log(@volatile var dir: File,
         // cause OffsetOutOfRangeException. To solve that, we cap the reading up to exposed position instead of the log
         // end of the active segment.
         val maxPosition = {
-          if (segmentEntry == segments.lastEntry) {
+          if (maxOffsetMetadata.segmentBaseOffset == segment.baseOffset) {
+            maxOffsetMetadata.relativePositionInSegment
+          } else if (segmentEntry == segments.lastEntry) {
```

Comment:
We probably can simplify the logic a bit. Since at this point, startOffset <= maxOffsetMetadata. We can use maxOffsetMetadata.relativePositionInSegment if maxOffsetMetadata is on the current segment. Otherwise, we can just use segment.size.